File: frontend/src/api/budget.ts
import { apiRequest } from "./client";

export type OptionItem = { id: number; name: string };
export type Income = {
  income_id: number;
  description_income: string;
  amount: string;
  income_date: string;
  income_type: string;
};
export type Allocation = {
  allocation_id: number;
  description_allocation: string;
  amount: string;
  allocation_date: string;
  allocation_group: string;
  allocation_target: string;
  notes: string | null;
};
export type Expense = {
  expense_id: number;
  description_expense: string;
  price: string;
  expense_date: string;
  subcategory_id: number;
  subcategory: string;
  category: string;
  payment_method_id: number;
  payment_method: string;
};

export type IncomePayload = Omit<Income, "income_id">;
export type AllocationPayload = Omit<Allocation, "allocation_id">;

export type ExpensePayload = {
  description_expense: string;
  price: string;
  expense_date: string;
  subcategory_id: number;
  payment_method_id: number;
};

function withYears(path: string, years: number[] = []) {
  const params = new URLSearchParams();
  years.forEach((year) => params.append("years", String(year)));
  const query = params.toString();
  return `${path}${query ? `?${query}` : ""}`;
}

export const getIncomes = (years: number[] = []) => apiRequest<Income[]>(withYears("/budget/incomes", years));
export const getAllocations = (years: number[] = []) => apiRequest<Allocation[]>(withYears("/budget/allocations", years));
export const getExpenses = (years: number[] = []) => apiRequest<Expense[]>(withYears("/budget/expenses", years));
export const getSubcategories = () => apiRequest<OptionItem[]>("/budget/options/subcategories");
export const getPaymentMethods = () => apiRequest<OptionItem[]>("/budget/options/payment-methods");
export const getBudgetYears = () => apiRequest<number[]>("/budget/years");

export const createIncome = (payload: IncomePayload) =>
  apiRequest<Income>("/budget/incomes", { method: "POST", body: JSON.stringify(payload) });
export const createAllocation = (payload: AllocationPayload) =>
  apiRequest<Allocation>("/budget/allocations", { method: "POST", body: JSON.stringify(payload) });
export const updateIncome = (id: number, payload: Partial<IncomePayload>) =>
  apiRequest<Income>(`/budget/incomes/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
export const updateAllocation = (id: number, payload: Partial<AllocationPayload>) =>
  apiRequest<Allocation>(`/budget/allocations/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
export const deleteIncome = (id: number) => apiRequest<void>(`/budget/incomes/${id}`, { method: "DELETE" });
export const deleteAllocation = (id: number) => apiRequest<void>(`/budget/allocations/${id}`, { method: "DELETE" });

export const createExpense = (payload: ExpensePayload) =>
  apiRequest<Expense>("/budget/expenses", { method: "POST", body: JSON.stringify(payload) });
export const updateExpense = (id: number, payload: Partial<ExpensePayload>) =>
  apiRequest<Expense>(`/budget/expenses/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
export const deleteExpense = (id: number) => apiRequest<void>(`/budget/expenses/${id}`, { method: "DELETE" });

File: backend/app/schemas/budget.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Field


class OptionItem(BaseModel):
    id: int
    name: str


class IncomeRead(BaseModel):
    income_id: int
    description_income: str
    amount: Decimal
    income_date: date
    income_type: str


class IncomeCreate(BaseModel):
    description_income: str = Field(min_length=1)
    amount: Decimal = Field(ge=0)
    income_date: date
    income_type: str = "Autre"


class IncomeUpdate(BaseModel):
    description_income: str | None = None
    amount: Decimal | None = Field(default=None, ge=0)
    income_date: date | None = None
    income_type: str | None = None


class AllocationRead(BaseModel):
    allocation_id: int
    description_allocation: str
    amount: Decimal
    allocation_date: date
    allocation_group: str
    allocation_target: str
    notes: str | None = None


class AllocationCreate(BaseModel):
    description_allocation: str = Field(min_length=1)
    amount: Decimal = Field(ge=0)
    allocation_date: date
    allocation_group: str = "Investissement"
    allocation_target: str = "Bourse"
    notes: str | None = None


class AllocationUpdate(BaseModel):
    description_allocation: str | None = None
    amount: Decimal | None = Field(default=None, ge=0)
    allocation_date: date | None = None
    allocation_group: str | None = None
    allocation_target: str | None = None
    notes: str | None = None


class ExpenseRead(BaseModel):
    expense_id: int
    description_expense: str
    price: Decimal
    expense_date: date
    subcategory_id: int
    subcategory: str
    category: str
    payment_method_id: int
    payment_method: str


class ExpenseCreate(BaseModel):
    description_expense: str = Field(min_length=1)
    price: Decimal = Field(ge=0)
    expense_date: date
    subcategory_id: int
    payment_method_id: int


class ExpenseUpdate(BaseModel):
    description_expense: str | None = None
    price: Decimal | None = Field(default=None, ge=0)
    expense_date: date | None = None
    subcategory_id: int | None = None
    payment_method_id: int | None = None
